src/popup_data.js

```diff
@@ -1,29 +1,62 @@
-import React from 'react';
-import * as ReactDOMServer from 'react-dom/server';
+import React from "react";
+import * as ReactDOMServer from "react-dom/server";
 
-
-function getContentAsString(data){
-  return Array.isArray(data)? data.join(', ') : data
+function getContentAsString(data) {
+  return Array.isArray(data) ? data.join(", ") : data;
 }
 
-function getFormattedDataForPopup(data){
-  return Object.keys(data).map(x => (
-    <p className="m-0">
-      <b>{x}</b>
-      {": " + getContentAsString(data[x])}
-    </p>));
+const mapCustomTypeToReactComponent = customValue => {
+  if (!customValue.type || !customValue.value) {
+    throw new Error("Custom value must have type and value properties");
+  }
+
+  const valueToDisplay = customValue?.displayValue || customValue.value;
+
+  switch (customValue.type) {
+    case "hyperlink":
+      return (
+        <a href={customValue.value} target="_blank">
+          {valueToDisplay}
+        </a>
+      );
+    default:
+      return getContentAsString(valueToDisplay);
+  }
+};
+
+const isCustomValue = value => typeof value === "object" && !(value instanceof Array);
+
+function getFormattedDataForPopup(data) {
+  return Object.entries(data).map(([dataKey, value]) => {
+    if (isCustomValue(value)) {
+      const CustomDataComponent = mapCustomTypeToReactComponent(value);
+
+      return (
+        <p key={dataKey} className="m-0">
+          <b>{dataKey}</b>
+          {": "}
+          {CustomDataComponent}
+        </p>
+      );
+    }
+
+    return (
+      <p key={dataKey} className="m-0">
+        <b>{dataKey}</b>
+        {": " + getContentAsString(value)}
+      </p>
+    );
```

Comment:
In this case - u can create variable which stores CustomDataComponent or getContentAsString(value) depends on condition and then pass this variable to the jsx, so you won't duplicate it

Reply:
```suggestion
    const DataComponent = isCustomValue(value) ? mapCustomTypeToReactComponent(value) : getContentAsString(value)
  
    return (
      <p key={dataKey} className="m-0">
        <b>{dataKey}</b>
        {": "}
        {DataComponent}
      </p>
    );
  };

```

Reply:
@KrystianKjjk sth like that, right?

Reply:
I have already refactored it to separate files in https://github.com/Problematy/goodmap-frontend/issues/18 as it's branch is based on this branch.

Reply:
Improved in [PR](https://github.com/Problematy/goodmap-frontend/pull/25/files#diff-ae78d83a350f971de8f0339822120adae40a4a033e98e400d7b1c62d0d80038e)